src/external/visualization_integration.cu

```diff
@@ -29,7 +31,49 @@ __host__ std::string vector_double_to_json_array(const std::vector<double> &v)
     return ans;
 }
 
-__host__ bool send_particles_to_visualization(const std::string &url, MapNode *nodes, int n_of_nodes)
+__host__ bool send_model_to_visualization(const std::pair<std::string, std::string> &urls, Polyhedron *polyhedron)
+{
+    std::vector<double> polyhedron_vertices, polyhedron_faces;
+    std::vector<std::vector<double>> poly;
+
+    for(int i = 0; i < polyhedron->get_n_of_faces(); ++i)
+    {
+        for(int j = 0; j < polyhedron->get_faces()[i].get_n_of_vertices(); ++j)
+        {
+            SpacePoint v = polyhedron->get_faces()[i].get_vertices()[j];
+            poly.push_back({v.x, v.y, v.z});
+        }
+    }
+
+    std::string body = "[ ";
+    for(int i = 0; i < poly.size(); ++i)
+    {
+        body += vector_double_to_json_array(poly[i]);
+        if (i != poly.size() - 1)
+            body += ',';
+    }
+    body += " ]";
```

Comment:
This is still a duplication: there is a very similar function (`vector_double_to_json_array`) that does the same thing just with a different array type. Moreover, absolutely the same logic is written in a slightly different way in that function (for example, you have spaces near square brackets while I don't; you use a condition inside of the loop to check if you have to add a comma while I rewrite the last comma with a bracket), so it can cause misunderstanding about the reasons of using different algorithms in different places. That's not what should happen

I would recommend you to think about how you would solve the problem before reading the rest of this comment.

Here is how I would implement this. I would add a template function with a name like `to_string_extended` accepting one parameter and returning an `std::string`. It would have two declarations:
`template<class T> std::string to_string_extended(const T &v)` would just return `std::to_string(v)`
and
`template<class T> std::string to_string_extended(const vector<T> &v)` would return a json array string where each element's string representation is `to_string_extended(v[i])`.
Then both all the calls to `vector_double_to_json_array` and this piece of code can be replaced with calls to `to_string_extended`

I hope, this is clear enough to code it, feel free to ask me about any details of this problem and/or my implementation idea and/or your own idea (i. e. if I would have accepted a different solution you thought of)